rfcs/0003-template-compiler.md

```diff
@@ -0,0 +1,126 @@
+// ...
+import { _$LH as litHtmlPrivate_1 } from "lit-html/private-ssr-support.js";
+const { AttributePart: _$LH_AttributePart, PropertyPart: _$LH_PropertyPart, BooleanAttributePart: _$LH_BooleanAttributePart, EventPart: _$LH_EventPart } = litHtmlPrivate_1;
+```
+
+## Implementation Considerations
+
+### Implementation Plan
+
+Create a new `@lit-labs/compiler` package which exports a Typescript transformer.
+
+### Backward Compatibility
+
+lit-html supports compiled templates on and above version 2.7.5.
+
+### Testing Plan
+
+Unit tests are sufficient for testing that the transform is applied correctly. An additional test target `test:dev:compiled` will be added to lit-html to compile the test suite using the transformer ensuring that most lit-html tests behave the same with a `CompiledTemplate`.
+
+> Note: Some lit-html tests need to be skipped, for example dev mode warnings that will no longer execute if the template has been compiled.
+
+### Performance and Code Size Impact
+
+This RFC has been prototyped: https://github.com/lit/lit/pull/3984 to gather benchmarks. Note that these are all subject to change based on the final implementation. These initial benchmarks support that compiling templates do result in a measurable performance improvement, with greater performance improvements occurring in template heavy contexts.
+
+**Kitchen sink** average tachometer benchmark results:
+
+- [**6.7% faster**]: `render-compiled`: 7.93ms - 8.10ms, `render-uncompiled`: 8.50ms - 8.69ms
+- [roughly same]: `update-compiled`: 17.92ms - 18.56ms, `update-uncompiled`: 18.08ms - 19.00ms
+- [roughly same]: `nop-update-compiled`: 5.53ms - 5.66ms, `nop-update-uncompiled`: 5.64ms - 5.80ms
+ - [**9.3% larger gzipped file size**]: `index.js` uncompiled is 2216 bytes gzipped, or compiled it is 2443 bytes gzipped.
+
+**Template Heavy** average tachometer benchmark results:
+
+ - [**46% faster**]: `render-compiled`: 9.51ms - 9.67ms, `render-uncompiled`: 17.66ms - 18.03ms
+ - [**21% faster**]: `update-compiled`: 26.40 - 26.91ms, `update-uncompiled`: 33.62 - 34.51ms
+ - [**5.6% larger gzipped file size**]: `template-heavy.js` uncompiled is 202,544 bytes gzipped, or compiled it is 214,495 bytes gzipped.
+
+**repeat** average tachometer benchmark results:
+
+ - [**14% faster**]: `render-compiled`: 7.40ms - 7.62ms, `render-uncompiled`: 8.56ms - 8.94ms
+ - [**8% faster**]: `update-compiled`: 189.57ms - 197.54ms, `update-uncompiled`: 206.51ms - 218.68ms
+ - [**7.1% larger gzipped file size**]: `index.js` uncompiled is 3072 bytes gzipped, or compiled it is 3307 bytes gzipped.
+
+### Interoperability
+
+This labs package is not interoperable as it specifically optimizes Lit template results using specific knowledge of the lit-html internal render phases.
+
+### Security Impact
+
+The prepared compiled template must be a tagged template literal to defend against JSON injection attacks. Otherwise the `@lit-labs/compiler` package is only a transform and has no security impact.
```

Comment:
More specifically, a compiled template is processed in the same way that an uncompiled template is, we're just moving some of the work to build time. This should preserve the security invariants of the code being compiled

Reply:
Much better wording - integrated.